### ISSUE TITLE
DNS response is not discarded when DNSSEC validation fails

### ISSUE THREAD

**Issue by Laudian** [systemd/systemd#34217]
### systemd version the issue has been seen with

255.4-1ubuntu8.4

### Used distribution

Ubuntu 24.04.1

### Linux kernel version used

6.8.0-41-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

Per RFC 4035 5.5, resolvers MUST NOT include the full DNS response if the response fails DNSSEC validation, unless the CD bit was set in the initial query.

### Unexpected behaviour you saw

However, I recently noticed that Systemd-Resolved does include the full response on responses that fail DNSSEC. This breaks behaviour in other applications. For example, `dig +short` does not check the status of the response but relies on the body being empty.

### Steps to reproduce the problem

Resolving with Systemd-Resolved (short):
```
dig +short example.com
104.21.51.142
172.67.181.103
```
Resolving with Systemd-Resolved (full-length):
```
dig example.com

; <<>> DiG 9.18.28-0ubuntu0.24.04.1-Ubuntu <<>> example.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 716
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;example.com.                    IN      A

;; ANSWER SECTION:
example.com.             300     IN      A       104.21.51.142
example.com.             300     IN      A       172.67.181.103

;; Query time: 99 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Sun Sep 01 17:46:11 CEST 2024
;; MSG SIZE  rcvd: 71
```
Resolving with other resolvers (here 1.1.1.1) (full-length):
```
dig example.com @1.1.1.1

; <<>> DiG 9.18.28-0ubuntu0.24.04.1-Ubuntu <<>> example.com @1.1.1.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 56984
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 9 (DNSKEY Missing): (no SEP matching the DS found for example.com.)
;; QUESTION SECTION:
;example.com.                    IN      A

;; Query time: 15 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Sun Sep 01 17:48:26 CEST 2024
;; MSG SIZE  rcvd: 89
```

**Comment by yuwata** [comment]
Hm? In my case, 1.1.1.1 also provides answer.
```
$ dig example.com @1.1.1.1

; <<>> DiG 9.18.28 <<>> example.com @1.1.1.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 46070
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;example.com.			IN	A

;; ANSWER SECTION:
example.com.		371	IN	A	93.184.215.14

;; Query time: 12 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Mon Sep 02 09:42:05 JST 2024
;; MSG SIZE  rcvd: 56
```
And, systemd-resolved provides the same.
```
$ dig example.com

; <<>> DiG 9.18.28 <<>> example.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 46943
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;example.com.			IN	A

;; ANSWER SECTION:
example.com.		904	IN	A	93.184.215.14

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Mon Sep 02 09:42:43 JST 2024
;; MSG SIZE  rcvd: 56
```

Could you provide the result of `resolvectl` ?

**Comment by Laudian** [comment]
Sorry, my bad. `example.com` was used as a placeholder for the real domain. It's not my domain and the owner has since fixed the DNSSEC configuration anyway, so sharing it wouldn't do any good. But I should obviously have mentioned this.

The response you received was NOERROR instead of SERVFAIL, so nothing was supposed to be discarded.

I'm afraid I don't have a domain where I could break the DNSSEC configuration right now for testing.

**Comment by yuwata** [comment]
I am not familiar, but maybe dnssec-failed.org can be used?

**Comment by Laudian** [comment]
Resolving with Systemd-Resolved (full):
```
dig dnssec-failed.org

; <<>> DiG 9.18.28-0ubuntu0.24.04.1-Ubuntu <<>> dnssec-failed.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 57631
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;dnssec-failed.org.             IN      A

;; ANSWER SECTION:
dnssec-failed.org.      300     IN      A       96.99.227.255

;; Query time: 928 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Mon Sep 02 11:17:49 CEST 2024
;; MSG SIZE  rcvd: 62
```
Resolving with Systemd-Resolved (short:
```
dig +short dnssec-failed.org
96.99.227.255
```
Resolving with 1.1.1.1 (full):
```
dig dnssec-failed.org @1.1.1.1

; <<>> DiG 9.18.28-0ubuntu0.24.04.1-Ubuntu <<>> dnssec-failed.org @1.1.1.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 2387
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 9 (DNSKEY Missing): (no SEP matching the DS found for dnssec-failed.org.)
;; QUESTION SECTION:
;dnssec-failed.org.             IN      A

;; Query time: 2 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Mon Sep 02 11:19:35 CEST 2024
;; MSG SIZE  rcvd: 103
```
Resolving with 1.1.1.1 (short):
```
dig +short dnssec-failed.org @1.1.1.1

```

Output of resolvectl:
```
Global
         Protocols: -LLMNR -mDNS -DNSOverTLS DNSSEC=yes/supported
  resolv.conf mode: stub

Link 2 (enp35s0)
    Current Scopes: DNS
         Protocols: +DefaultRoute -LLMNR -mDNS -DNSOverTLS DNSSEC=yes/supported
Current DNS Server: 2a01:4f8:0:1::add:9999
       DNS Servers: 213.133.98.98 213.133.100.100 213.133.99.99
                    2a01:4f8:0:1::add:9898 2a01:4f8:0:1::add:9999
                    2a01:4f8:0:1::add:1010

Link 3 (tun0)
    Current Scopes: none
         Protocols: -DefaultRoute -LLMNR -mDNS -DNSOverTLS DNSSEC=yes/supported

Link 4 (docker0)
    Current Scopes: none
         Protocols: -DefaultRoute -LLMNR -mDNS -DNSOverTLS DNSSEC=yes/supported

Link 5 (br-b57e6468a759)
    Current Scopes: none
         Protocols: -DefaultRoute -LLMNR -mDNS -DNSOverTLS DNSSEC=yes/supported
```

**Comment by rpigott** [comment]
This was addressed in #31952 and followups (v256+).

**Comment by yuwata** [comment]
OK. Then, let's close.
